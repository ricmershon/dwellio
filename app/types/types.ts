import { StructuredFormErrorMap } from "@/app/utils/build-form-error-map";

export interface Rates {
    nightly?: number;
    weekly?: number;
    monthly?: number
}

export type ActionState = {
    message?: string | null;
    status?: 'SUCCESS' | 'ERROR' | null,
    isFavorite?: boolean,
    isRead?: boolean,
    formData?: FormData,
    formErrorMap?: StructuredFormErrorMap
}

export interface PropertiesQuery {
    $or: [
        { name: RegExp },
        { description: RegExp },
        { amenities: RegExp },
        { type: RegExp },
        { 'location.street': RegExp },
        { 'location.city': RegExp },
        { 'location.state': RegExp },
        { 'location.zip': RegExp }
    ]
}

export interface PropertyImageData {
    secureUrl: string;
    publicId: string;
}

export type FormErrorsType = Record<string, string[]> | string[];

export const MAX_ITEMS_PER_PAGE = 6;