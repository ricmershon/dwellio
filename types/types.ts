import { StructuredFormErrorMap } from "@/utils/build-form-error-map";

export const ActionStatus = {
    INFO: 'info',
    SUCCESS: 'success',
    WARNING: 'warning',
    ERROR: 'error',
} as const;

type ActionStatus = (typeof ActionStatus)[keyof typeof ActionStatus];

export type ActionState = {
  message?: string | null;
  status?: ActionStatus | null;
  isFavorite?: boolean;
  isRead?: boolean;
  formData?: FormData;
  formErrorMap?: StructuredFormErrorMap;
};

export interface PropertiesQuery {
    $or: [
        { name: RegExp },
        { description: RegExp },
        { amenities: RegExp },
        { type: RegExp },
        { 'location.street': RegExp },
        { 'location.city': RegExp },
        { 'location.state': RegExp },
        { 'location.zip': RegExp }
    ]
}

export interface PropertyImageData {
    secureUrl: string;
    publicId: string;
    width: number;
    height: number;
}

export interface Rates {
    nightly?: number;
    weekly?: number;
    monthly?: number
}

export interface Amenity {
    id: string,
    value: string
}

export interface OptionType {
    value: string;
    label: string;
}

export type FormErrorsType = Record<string, string[]> | string[];

export const VIEWPORT_WIDTH_COOKIE_NAME = 'viewport_width';