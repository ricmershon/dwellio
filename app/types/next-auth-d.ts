// eslint-disable-next-line @typescript-eslint/no-unused-vars
import NextAuth, { DefaultSession } from "next-auth";

declare module 'next-auth' {
    interface Profile {
        picture?: string;
    }

    interface Session {
        user: {
            id?: string;
        } & DefaultSession["user"];
    }
}