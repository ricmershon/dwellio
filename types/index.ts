import { StructuredFormErrorMap } from "@/utils/build-form-error-map";

export const ActionStatus = {
    INFO: "info",
    SUCCESS: "success",
    WARNING: "warning",
    ERROR: "error",
} as const;

type ActionStatus = (typeof ActionStatus)[keyof typeof ActionStatus];

export type ActionState = {
    message?: string | null;
    status?: ActionStatus | null;
    isFavorite?: boolean;
    isRead?: boolean;
    formData?: FormData;
    formErrorMap?: StructuredFormErrorMap;
    userId?: string;
    isAccountLinked?: boolean;
    canSignInWith?: string[];
    email?: string;
    password?: string;
    shouldAutoLogin?: boolean;
};

export interface PropertiesQuery {
    $or: [
        { name: RegExp },
        { description: RegExp },
        { amenities: RegExp },
        { type: RegExp },
        { "location.street": RegExp },
        { "location.city": RegExp },
        { "location.state": RegExp },
        { "location.zip": RegExp }
    ]
}

export interface PropertyImageData {
    secureUrl: string;
    publicId: string;
    width: number;
    height: number;
}

export interface Rates {
    nightly?: number;
    weekly?: number;
    monthly?: number
}

export interface Amenity {
    id: string,
    value: string
}

export interface OptionType {
    value: string;
    label: string;
}

export type FormErrorsType = Record<string, string[]> | string[];

export interface AutocompleteFetchOptions {
    signal?: AbortSignal;
    includeDetails?: boolean;
    maxDetails?: number;
    regionCode?: string;
    includedPrimaryTypes?: string[];
};

export interface AutocompleteSuggestion {
    placePrediction: {
        placeId: string;
        text?: {
            text: string;
            languageCode?: string;
        };
        structuredFormat?: {
            mainText?: { text: string; matches?: Array<{ startOffset: number; endOffset: number }> };
            secondaryText?: { text: string; matches?: Array<{ startOffset: number; endOffset: number }> };
        };
    };
}

export interface AutocompletePrediction {
    placeId: string;
    text: string;
    structuredFormat?: AutocompleteSuggestion["placePrediction"]["structuredFormat"];
    street?: string | null;
    city?: string | null;
    state?: string | null;
    zipcode?: string | null;
}

export interface AutocompleteResponse {
    suggestions?: AutocompleteSuggestion[];
}

export interface AddressComponent {
    types?: string[];
    longText?: string;
    shortText?: string;
};

export interface PasswordValidation {
    isValid: boolean;
    errors: string[];
}

export const VIEWPORT_WIDTH_COOKIE_NAME = "viewport_width";