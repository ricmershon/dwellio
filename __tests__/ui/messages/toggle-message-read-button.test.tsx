import React from 'react';
import { render, screen, fireEvent, waitFor, createReactToastifyMock } from '@/__tests__/test-utils';

import ToggleMessageReadButton from '@/ui/messages/toggle-message-read-button';
import { ActionStatus } from '@/types/types';

// Mock external dependencies
jest.mock('@/lib/actions/message-actions', () => ({
    toggleMessageRead: jest.fn()
}));

jest.mock('@/context/global-context', () => ({
    useGlobalContext: jest.fn()
}));

jest.mock('react-toastify', () => createReactToastifyMock());

// Import mocked functions
const { toggleMessageRead } = jest.requireMock('@/lib/actions/message-actions');
const { useGlobalContext } = jest.requireMock('@/context/global-context');
const { toast } = jest.requireMock('react-toastify');

describe('ToggleMessageReadButton', () => {
    const mockSetUnreadCount = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
        
        useGlobalContext.mockReturnValue({
            unreadCount: 5,
            setUnreadCount: mockSetUnreadCount,
            staticInputs: {},
            setStaticInputs: jest.fn()
        });
    });

    describe('State Management', () => {
        it('should initialize isRead state from props', () => {
            render(<ToggleMessageReadButton messageId="msg-123" read={false} />);
            
            const button = screen.getByRole('button');
            expect(button).toHaveTextContent('Mark as Read');
        });

        it('should initialize with read state true', () => {
            render(<ToggleMessageReadButton messageId="msg-456" read={true} />);
            
            const button = screen.getByRole('button');
            expect(button).toHaveTextContent('Mark as Unread');
        });

        it('should update local state when action succeeds', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Message marked read.',
                isRead: true
            });

            render(<ToggleMessageReadButton messageId="msg-123" read={false} />);
            
            const button = screen.getByRole('button');
            expect(button).toHaveTextContent('Mark as Read');
            
            fireEvent.click(button);
            
            await waitFor(() => {
                expect(button).toHaveTextContent('Mark as Unread');
            });
        });

        it('should sync state with server response isRead flag', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Message marked unread.',
                isRead: false
            });

            render(<ToggleMessageReadButton messageId="msg-456" read={true} />);
            
            const button = screen.getByRole('button');
            fireEvent.click(button);
            
            await waitFor(() => {
                expect(button).toHaveTextContent('Mark as Read');
            });
        });

        it('should maintain state consistency across re-renders', () => {
            const { rerender } = render(<ToggleMessageReadButton messageId="msg-123" read={false} />);
            
            const button = screen.getByRole('button');
            expect(button).toHaveTextContent('Mark as Read');
            
            rerender(<ToggleMessageReadButton messageId="msg-123" read={false} />);
            
            expect(button).toHaveTextContent('Mark as Read');
        });
    });

    describe('User Interactions', () => {
        it('should handle form submission on button click', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Success',
                isRead: true
            });

            render(<ToggleMessageReadButton messageId="msg-789" read={false} />);
            
            const button = screen.getByRole('button');
            fireEvent.click(button);
            
            await waitFor(() => {
                expect(toggleMessageRead).toHaveBeenCalledWith('msg-789');
            });
        });

        it('should call toggleMessageRead server action', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Success',
                isRead: true
            });

            render(<ToggleMessageReadButton messageId="test-msg-id" read={false} />);
            
            fireEvent.click(screen.getByRole('button'));
            
            await waitFor(() => {
                expect(toggleMessageRead).toHaveBeenCalledTimes(1);
                expect(toggleMessageRead).toHaveBeenCalledWith('test-msg-id');
            });
        });

        it('should pass correct messageId to action', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Success',
                isRead: false
            });

            render(<ToggleMessageReadButton messageId="unique-message-id" read={true} />);
            
            fireEvent.click(screen.getByRole('button'));
            
            await waitFor(() => {
                expect(toggleMessageRead).toHaveBeenCalledWith('unique-message-id');
            });
        });

        it('should handle multiple sequential clicks', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Success',
                isRead: true
            });

            render(<ToggleMessageReadButton messageId="msg-123" read={false} />);
            
            const button = screen.getByRole('button');
            
            // First click
            fireEvent.click(button);
            
            await waitFor(() => {
                expect(toggleMessageRead).toHaveBeenCalledTimes(1);
            });

            // Should allow subsequent clicks
            expect(() => fireEvent.click(button)).not.toThrow();
        });
    });

    describe('Global Context Integration', () => {
        it('should call setUnreadCount on successful toggle to read', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Message marked read.',
                isRead: true
            });

            render(<ToggleMessageReadButton messageId="msg-123" read={false} />);
            
            fireEvent.click(screen.getByRole('button'));
            
            await waitFor(() => {
                expect(mockSetUnreadCount).toHaveBeenCalledTimes(1);
                expect(mockSetUnreadCount).toHaveBeenCalledWith(expect.any(Function));
            });

            // Test the function passed to setUnreadCount
            const updateFunction = mockSetUnreadCount.mock.calls[0][0];
            expect(updateFunction(10)).toBe(9); // Should decrement by 1
        });

        it('should increment count when marking as unread', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Message marked unread.',
                isRead: false
            });

            render(<ToggleMessageReadButton messageId="msg-456" read={true} />);
            
            fireEvent.click(screen.getByRole('button'));
            
            await waitFor(() => {
                expect(mockSetUnreadCount).toHaveBeenCalledWith(expect.any(Function));
            });

            // Test the function passed to setUnreadCount
            const updateFunction = mockSetUnreadCount.mock.calls[0][0];
            expect(updateFunction(5)).toBe(6); // Should increment by 1
        });

        it('should decrement count when marking as read', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Message marked read.',
                isRead: true
            });

            render(<ToggleMessageReadButton messageId="msg-789" read={false} />);
            
            fireEvent.click(screen.getByRole('button'));
            
            await waitFor(() => {
                expect(mockSetUnreadCount).toHaveBeenCalledWith(expect.any(Function));
            });

            // Test the function passed to setUnreadCount
            const updateFunction = mockSetUnreadCount.mock.calls[0][0];
            expect(updateFunction(8)).toBe(7); // Should decrement by 1
        });

        it('should not update count on action failure', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.ERROR,
                message: 'Failed to update message'
            });

            render(<ToggleMessageReadButton messageId="msg-123" read={false} />);
            
            fireEvent.click(screen.getByRole('button'));
            
            await waitFor(() => {
                expect(toggleMessageRead).toHaveBeenCalled();
            });

            expect(mockSetUnreadCount).not.toHaveBeenCalled();
        });
    });

    describe('Toast Notifications', () => {
        it('should not show toast on successful actions', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Message updated successfully'
            });

            render(<ToggleMessageReadButton messageId="msg-123" read={false} />);
            
            fireEvent.click(screen.getByRole('button'));
            
            await waitFor(() => {
                expect(toggleMessageRead).toHaveBeenCalled();
            });

            expect(toast.error).not.toHaveBeenCalled();
        });

        it('should show error toast for failed actions', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.ERROR,
                message: 'Failed to toggle message status'
            });

            render(<ToggleMessageReadButton messageId="msg-456" read={true} />);
            
            fireEvent.click(screen.getByRole('button'));
            
            await waitFor(() => {
                expect(toast.error).toHaveBeenCalledWith('Failed to toggle message status');
            });
        });

        it('should handle ActionStatus.SUCCESS responses', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Message marked as read',
                isRead: true
            });

            render(<ToggleMessageReadButton messageId="msg-123" read={false} />);
            
            fireEvent.click(screen.getByRole('button'));
            
            await waitFor(() => {
                expect(toggleMessageRead).toHaveBeenCalled();
            });

            expect(toast.error).not.toHaveBeenCalled();
        });

        it('should handle ActionStatus.ERROR responses', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.ERROR,
                message: 'Database connection failed'
            });

            render(<ToggleMessageReadButton messageId="msg-789" read={false} />);
            
            fireEvent.click(screen.getByRole('button'));
            
            await waitFor(() => {
                expect(toast.error).toHaveBeenCalledWith('Database connection failed');
            });
        });
    });

    describe('Button Text Logic', () => {
        it('should show "Mark as Read" for unread messages', () => {
            render(<ToggleMessageReadButton messageId="msg-123" read={false} />);
            
            const button = screen.getByRole('button');
            expect(button).toHaveTextContent('Mark as Read');
        });

        it('should show "Mark as Unread" for read messages', () => {
            render(<ToggleMessageReadButton messageId="msg-456" read={true} />);
            
            const button = screen.getByRole('button');
            expect(button).toHaveTextContent('Mark as Unread');
        });

        it('should update button text after successful toggle', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Success',
                isRead: true
            });

            render(<ToggleMessageReadButton messageId="msg-123" read={false} />);
            
            const button = screen.getByRole('button');
            expect(button).toHaveTextContent('Mark as Read');
            
            fireEvent.click(button);
            
            await waitFor(() => {
                expect(button).toHaveTextContent('Mark as Unread');
            });
        });

        it('should maintain text consistency with state', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Success',
                isRead: false
            });

            render(<ToggleMessageReadButton messageId="msg-456" read={true} />);
            
            const button = screen.getByRole('button');
            fireEvent.click(button);
            
            await waitFor(() => {
                expect(button).toHaveTextContent('Mark as Read');
            });
        });
    });

    describe('Error Handling', () => {
        it('should handle server action failures gracefully', async () => {
            toggleMessageRead.mockRejectedValue(new Error('Network error'));

            render(<ToggleMessageReadButton messageId="msg-123" read={false} />);
            
            const button = screen.getByRole('button');
            
            // Should not crash on error
            expect(() => fireEvent.click(button)).not.toThrow();
        });

        it('should maintain original state on error', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.ERROR,
                message: 'Update failed'
            });

            render(<ToggleMessageReadButton messageId="msg-123" read={false} />);
            
            const button = screen.getByRole('button');
            expect(button).toHaveTextContent('Mark as Read');
            
            fireEvent.click(button);
            
            await waitFor(() => {
                expect(toast.error).toHaveBeenCalled();
            });

            // Button text should remain the same
            expect(button).toHaveTextContent('Mark as Read');
        });

        it('should display error messages to user', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.ERROR,
                message: 'Authorization failed'
            });

            render(<ToggleMessageReadButton messageId="msg-456" read={true} />);
            
            fireEvent.click(screen.getByRole('button'));
            
            await waitFor(() => {
                expect(toast.error).toHaveBeenCalledWith('Authorization failed');
            });
        });

        it('should not update global context on error', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.ERROR,
                message: 'Database error'
            });

            render(<ToggleMessageReadButton messageId="msg-789" read={false} />);
            
            fireEvent.click(screen.getByRole('button'));
            
            await waitFor(() => {
                expect(toast.error).toHaveBeenCalled();
            });

            expect(mockSetUnreadCount).not.toHaveBeenCalled();
        });
    });

    describe('Form Integration', () => {
        it('should wrap button in form element', () => {
            render(<ToggleMessageReadButton messageId="msg-123" read={false} />);
            
            const form = screen.getByRole('button').closest('form');
            expect(form).toBeInTheDocument();
        });

        it('should use server action as form action', () => {
            render(<ToggleMessageReadButton messageId="msg-456" read={true} />);
            
            const form = screen.getByRole('button').closest('form');
            expect(form).toHaveAttribute('action');
        });

        it('should handle form submission properly', async () => {
            toggleMessageRead.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                message: 'Success',
                isRead: true
            });

            render(<ToggleMessageReadButton messageId="msg-789" read={false} />);
            
            const button = screen.getByRole('button');
            expect(button).toHaveAttribute('type', 'submit');
            
            fireEvent.click(button);
            
            await waitFor(() => {
                expect(toggleMessageRead).toHaveBeenCalled();
            });
        });

        it('should apply correct CSS classes', () => {
            render(<ToggleMessageReadButton messageId="msg-123" read={false} />);
            
            const form = screen.getByRole('button').closest('form');
            expect(form).toHaveClass('inline-block');
            
            const button = screen.getByRole('button');
            expect(button).toHaveClass('btn', 'btn-primary', 'mt-4', 'mr-2');
        });
    });

    describe('Component Props', () => {
        it('should accept messageId prop', () => {
            render(<ToggleMessageReadButton messageId="test-id-123" read={false} />);
            
            const button = screen.getByRole('button');
            expect(button).toBeInTheDocument();
        });

        it('should accept read boolean prop', () => {
            render(<ToggleMessageReadButton messageId="msg-456" read={true} />);
            
            const button = screen.getByRole('button');
            expect(button).toHaveTextContent('Mark as Unread');
        });

        it('should handle different messageId formats', async () => {
            const testIds = ['msg-123', 'message_456', '507f1f77bcf86cd799439011'];
            
            for (const messageId of testIds) {
                toggleMessageRead.mockResolvedValue({
                    status: ActionStatus.SUCCESS,
                    message: 'Success',
                    isRead: true
                });

                const { unmount } = render(<ToggleMessageReadButton messageId={messageId} read={false} />);
                
                fireEvent.click(screen.getByRole('button'));
                
                await waitFor(() => {
                    expect(toggleMessageRead).toHaveBeenCalledWith(messageId);
                });

                // Clean up for next iteration
                unmount();
                jest.clearAllMocks();
            }
        });
    });
});