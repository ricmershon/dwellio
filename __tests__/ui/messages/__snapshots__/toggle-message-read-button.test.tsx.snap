// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ToggleMessageReadButton Snapshots should match snapshot for form structure: form-structure 1`] = `
<form
  action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
  class="inline-block"
>
  <button
    class="btn btn-primary mt-4 mr-2"
    type="submit"
  >
    Mark as Unread
  </button>
</form>
`;

exports[`ToggleMessageReadButton Snapshots should match snapshot for read message state (Mark as Unread): read-mark-as-unread 1`] = `
<form
  action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
  class="inline-block"
>
  <button
    class="btn btn-primary mt-4 mr-2"
    type="submit"
  >
    Mark as Unread
  </button>
</form>
`;

exports[`ToggleMessageReadButton Snapshots should match snapshot for unread message state (Mark as Read): unread-mark-as-read 1`] = `
<form
  action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
  class="inline-block"
>
  <button
    class="btn btn-primary mt-4 mr-2"
    type="submit"
  >
    Mark as Read
  </button>
</form>
`;

exports[`ToggleMessageReadButton Snapshots should match snapshot with different message ID: different-message-id 1`] = `
<form
  action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
  class="inline-block"
>
  <button
    class="btn btn-primary mt-4 mr-2"
    type="submit"
  >
    Mark as Read
  </button>
</form>
`;
